### PR TITLE
wgsl: Add `sparseMatrixRange` to FPTraits API

### LINKED ISSUE #3174: wgsl: Expose various test range generators in FPTraits
There are a variety of functions for generating test ranges that are effective parameterized by fp type (vectorF16Range vs vectorF32Range vs vectorF64). Because these are not part of the FPTraits there is a bunch of cases where case generation code is needing different functions instead of doing something like FP[trait].vectorRange.

dot for example was actually doing the wrong thing, and using F32 for both f32 and f16, so there are probably some bugs from this to be quashed.

This bug covers adding the APIs to FPTraits, and refactoring the existing code to use them.

### PR DESCRIPTION
**Pull request by zoddicus**
Fixes #3174

<hr>

**Requirements for PR author:**

- [ ] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [ ] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [ ] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Comment by zoddicus** [comment]
Another stacked change, all the new code is in https://github.com/gpuweb/cts/pull/3183/commits/744e31bd39b2d1f32468166431f3b3f92eb9ce44

File: src/webgpu/shader/execution/expression/call/builtin/transpose.spec.ts [hunk 1]
```diff
@@ -10,71 +10,37 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
-          sparseMatrixF64Range(cols, rows),
-          'finite',
-          FP.abstract.transposeInterval
-        );
-      },
-    }))
-  )
-  .reduce((a, b) => ({ ...a, ...b }), {});
-
 export const d = makeCaseCache('transpose', {
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/transpose.spec.ts:42]: Just
``` suggestion
export const d = makeCaseCache('transpose', cases);
```

**Review by jzm-intel — APPROVED** [review]
Looks good to me, thanks

File: src/webgpu/shader/execution/expression/call/builtin/atanh.spec.ts [hunk 2]
```diff
@@ -12,46 +12,39 @@ Note: The result is not mathematically meaningful when abs(e) >= 1.
...
-  ...fullF32Range(),
-];
-const f16_inputs = [
-  ...biasedRange(kValue.f16.negative.less_than_one, -0.9, 20), // discontinuity at x = -1
-  -1,
-  ...biasedRange(kValue.f16.positive.less_than_one, 0.9, 20), // discontinuity at x = 1
-  1,
-  ...fullF16Range(),
-];
+const cases = (['f32', 'f16'] as const)
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/expression/call/builtin/atanh.spec.ts:25]: For this and the other changes, a comment describing what `cases` holds (non-exhaustive fields) would be good as this made me go cross-eyed trying to figure out what it was actually doing

**Review by ben-clayton — APPROVED** [review]
LGTM, however I'm not sure if the cache should be re-generated or not based on the code changes. It might be worth checking whether you get the same diffs if we all agree on a NodeJS version to use. See https://github.com/gpuweb/cts/pull/3187

**Comment by zoddicus** [comment]
> LGTM, however I'm not sure if the cache should be re-generated or not based on the code changes. It might be worth checking whether you get the same diffs if we all agree on a NodeJS version to use. See #3187

I am currently using 20.9.0, which I believe is the newest LTS release and the same one that you pin in that PR.

This PR shouldn't be changing the actual expectations significantly (There are a couple of tests that were using the wrong generator that will be changing, but most should be a no-op). It does change the code in floating_point.ts and the various .spec.ts files, so I think they should have a different hash.

**Comment by ben-clayton** [comment]
SGTM then, thanks!

File: src/webgpu/shader/execution/expression/call/builtin/atanh.spec.ts (hunk 2 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/atanh.spec.ts:25]: Do you think something like
```
// Cases: [f32|f16]_[non_]const
```

is sufficient?

This is what jzm has included for the usage of this pattern that they wrote, but I didn't replicate consistantly.

File: src/webgpu/shader/execution/expression/call/builtin/transpose.spec.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/transpose.spec.ts]: Done

**Pull request merged**